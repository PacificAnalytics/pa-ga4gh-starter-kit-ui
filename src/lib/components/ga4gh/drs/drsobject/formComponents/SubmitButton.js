import React from 'react';
import {
    FormControl,
    Button
} from '@material-ui/core'
import {
    useHistory
} from 'react-router-dom';
import { SpaceDivider } from '../../../../common/layout';
import FormViewType from '../../../../../model/common/FormViewType';
import { scrollToTop } from '../../../../../functions/common';
import ApiCaller from '../../../../../utils/ApiCaller';

/*
    Used to make a POST or PUT requests to create or update a new DRS Object.
    If the request is successful, the user is redirected to the DRS Index
    page and a success message is displayed. However, if it is unsuccessful, 
    the user is not redirected and an error message is displayed.
*/
const SubmitButton = props => {

    const history = useHistory();

    const submit = () => {

        const validateActiveDrsObject = () => {
            let [isValid, message] = [true, ''];

            // validate Id
            if (props.activeDrsObject.id === '' || props.activeDrsObject.id === null) {
                return [false, `'Id' parameter must be set`];
            }

            // validate aliases
            props.activeDrsObject.aliases.forEach((alias, index) => {
                if (!alias) {
                    return [isValid, message] = [false, `Invalid 'Alias' at position: ${index}`];
                }
            })

            if(!props.activeDrsObject.is_bundle) {
                // validate checksums
                props.activeDrsObject.checksums.forEach((checksum, index) => {
                    if (!checksum.type || !checksum.checksum) {
                        return [isValid, message] = [false, `Invalid 'Checksum' at position ${index}`];
                    }
                })

                //validate file access objects
                props.activeDrsObject.file_access_objects.forEach((fileAccessObject, index) => {
                    if (!fileAccessObject.path) {
                        return [isValid, message] = [false, `Invalid 'Local File Access Point' at position: ${index}`];
                    }
                })

                //  validate AWS S3 access objects
                props.activeDrsObject.aws_s3_access_objects.forEach((awsS3AccessObject, index) => {
                    if (!awsS3AccessObject.region || !awsS3AccessObject.bucket || !awsS3AccessObject.key) {
                        return [isValid, message] = [false, `Invalid 'AWS S3 Access Point' at position ${index}`];
                    }
                })    
            }

            // validate children and parents arrays
            const validateRelativesArray = (relatives, name) => {
                relatives.forEach((relative, index) => {
                    if (relative.isValid !== true) {
                        [isValid, message] = [false, `Invalid ${name} at position: ${index}, id: '${relative.id}'`];
                        return;
                    }
                })
            }

            if(props.activeDrsObject.is_bundle) {
                validateRelativesArray(props.activeDrsObject.drs_object_children, 'child');
                if (!isValid) {
                    return [isValid, message];
                }    
            }

            validateRelativesArray(props.activeDrsObject.drs_object_parents, 'parent');
            if (!isValid) {
                return [isValid, message];
            }

            return [isValid, message];
        }

        // prepares the request body from the activeDrsObject
        const prepareRequestBody = () => {
            let requestBody = {...props.activeDrsObject};
            let blobOnlyParams = ['mime_type', 'size', 'checksums', 'file_access_objects', 'aws_s3_access_objects'];
            let bundleOnlyParams = ['drs_object_children'];
            let toUnsetIfBundle = {
                'true': blobOnlyParams,
                'false': bundleOnlyParams 
            }
            toUnsetIfBundle[requestBody.is_bundle].forEach(toUnset => delete requestBody[toUnset]);
            return requestBody;
        }

        const determineRequestConfig = requestBody => {
            let requestConfig = {};
            switch (props.formViewType) {
                case FormViewType.NEW:
                    requestConfig = {
                        url: `${props.adminURL}/admin/ga4gh/drs/v1/objects`,
                        method: 'POST'
                    }
                    break;
                case FormViewType.EDIT:
                    requestConfig = {
                        url: `${props.adminURL}/admin/ga4gh/drs/v1/objects/${props.activeDrsObject.id}`,
                        method: 'PUT'
                    }
                    break
            }
            requestConfig.data = requestBody;
            return requestConfig;
        }

        const determineSuccessMessage = id => {
            let successMessage = '';
            switch (props.formViewType) {
                case FormViewType.NEW:
                    successMessage = `Successfully created DrsObject with id: '${id}'`;
                break;
                case FormViewType.EDIT: 
                    successMessage = `Successfully updated DrsObject with id: '${id}'`;
                break;
            }
            return successMessage;
        }

        const executeApiCall = requestConfig => {
            ApiCaller(
                requestConfig,
                responseData => {
                    props.setSuccessMessage(determineSuccessMessage(responseData.id));
                    props.retrieveDrsObjectsList();
                    history.push(props.baseURL);
                    scrollToTop();
                },
                props.setError
            );
        }

        let [isValid, message] = validateActiveDrsObject();
        if (!isValid) {
            props.setError(new Error(message));
            return;
        }
        let requestBody = prepareRequestBody();
        let requestConfig = determineRequestConfig(requestBody);
        executeApiCall(requestConfig);
    }

    return (
        <FormControl fullWidth>
            <SpaceDivider/>
            <Button
                variant='contained'
                color='primary'
                onClick={submit}
            >
                Submit
            </Button>
        </FormControl>
    )
}

export default SubmitButton;
