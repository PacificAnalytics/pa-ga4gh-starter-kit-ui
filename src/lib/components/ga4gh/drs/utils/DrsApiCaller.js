import axios from 'axios';

const DrsApiCaller = (requestConfig, handleResponseData, handleError) => {
  axios(requestConfig)
  .then(response => {
    handleResponseData(response.data);
  })
  .catch(error => {
    handleError(error)
  })
}

export default DrsApiCaller;
