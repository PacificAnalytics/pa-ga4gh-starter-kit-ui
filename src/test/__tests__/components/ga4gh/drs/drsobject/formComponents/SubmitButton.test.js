import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import SubmitButton from '../../../../../lib/components/drs/formComponents/SubmitButton';
import {
    mockValidTestBlob, 
    mockValidTestBundle
} from '../../../../resources/MockData';
import {
    setSuccessMessage, 
    retrieveDrsObjectsList, 
    setError
} from '../../../../resources/MockFunctions';

jest.mock('../../../../../lib/components/drs/utils/DrsApiCaller');
import DrsApiCaller from '../../../../../lib/components/drs/utils/DrsApiCaller';
DrsApiCaller.mockImplementation(() => {
    setSuccessMessage(`Successfully submitted DrsObject with id: '${mockValidTestBlob.id}'`);
    retrieveDrsObjectsList();
});

let submitMockFunctions = null;

beforeEach(() => {
    submitMockFunctions = {
        setSuccessMessage: setSuccessMessage,
        retriveDrsObejctsList: retrieveDrsObjectsList,
        setError: setError,
    }
});

afterEach(() => {
    setSuccessMessage.mockClear();
    retrieveDrsObjectsList.mockClear();
    setError.mockClear();
});

test('<SubmitButton/> can handle submitting a valid blob', () => {
    let activeDrsObject = {...mockValidTestBlob};
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    let submitButton = screen.getByRole('button');
    expect(submitButton).toBeInTheDocument();
    userEvent.click(submitButton);
    expect(setSuccessMessage).toHaveBeenCalled();
    expect(retrieveDrsObjectsList).toHaveBeenCalled();
    expect(setError).not.toHaveBeenCalled();
});

test('<SubmitButton/> can handle submitting a valid bundle', () => {
    let activeDrsObject = {...mockValidTestBundle};
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    let submitButton = screen.getByRole('button');
    expect(submitButton).toBeInTheDocument();
    userEvent.click(submitButton);
    expect(setSuccessMessage).toHaveBeenCalled();
    expect(retrieveDrsObjectsList).toHaveBeenCalled();
    expect(setError).not.toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid id', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.id = '';
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid aliases', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.aliases = ["NewDrsAlias", ""];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid checksum type', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.checksums[0] = [{"checksum": "1234", "type": ""}];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid checksum value', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.checksums[0] = [{"checksum": "", "type": "md5"}];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid file access object', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.file_access_objects[0] = [
        {
            "path": ""
        }
    ];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid AWS S3 access object region', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.aws_s3_access_objects[0] = [
        {
            "region": "",
            "bucket": "test-bucket",
            "key": "test-key"
        }
    ];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid AWS S3 access object bucket', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.aws_s3_access_objects[0] = [
        {
            "region": "test-region",
            "bucket": "",
            "key": "test-key"
        }
    ];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle invalid AWS S3 access object key', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.aws_s3_access_objects[0] = [
        {
            "region": "test-region",
            "bucket": "test-bucket",
            "key": ""
        }
    ];
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle submitting a blob with an invalid parent', () => {
    let activeDrsObject = {...mockValidTestBlob};
    activeDrsObject.drs_object_parents[0].isValid = false;
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});

test('<SubmitButton/> can handle submitting a bundle with an invalid child', () => {
    let activeDrsObject = {...mockValidTestBundle};
    activeDrsObject.drs_object_children[0].isValid = false;
    let {container} = render(
        <SubmitButton 
            activeDrsObject={activeDrsObject}
            formViewType='NEW'
            {...submitMockFunctions}
        />
    );
    userEvent.click(screen.getByRole('button'));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});