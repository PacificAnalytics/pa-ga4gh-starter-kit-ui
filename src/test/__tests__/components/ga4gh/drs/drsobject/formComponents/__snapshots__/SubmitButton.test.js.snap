// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubmitButton/> can handle submitting a valid blob 1`] = `
<div
  class="MuiFormControl-root MuiFormControl-fullWidth"
>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <hr
      class="MuiDivider-root"
    />
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<SubmitButton/> can handle submitting a valid bundle 1`] = `
<div
  class="MuiFormControl-root MuiFormControl-fullWidth"
>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <hr
      class="MuiDivider-root"
    />
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
